io/io/src/TBufferMerger.cxx

```diff
@@ -90,37 +89,28 @@ void TBufferMerger::SetAutoSave(size_t size)
 
 void TBufferMerger::Merge()
 {
-   fBuffered = 0;
-   fMerger.PartialMerge();
-   fMerger.Reset();
-
-   if (fCallback)
-      fCallback();
-}
-
-void TBufferMerger::WriteOutputFile()
-{
-   std::unique_ptr<TBufferFile> buffer;
-
-   while (true) {
-      std::unique_lock<std::mutex> lock(fQueueMutex);
-      fDataAvailable.wait(lock, [this]() { return !this->fQueue.empty(); });
-
-      buffer.reset(fQueue.front());
-      fQueue.pop();
-      lock.unlock();
+   std::queue<TBufferFile *> queue;
+   {
+      std::lock_guard<std::mutex> q(fQueueMutex);
+      std::swap(queue, fQueue);
+      fBuffered = 0;
+   }
 
-      if (!buffer)
-         break;
+   std::vector<TMemFile*> files;
+   while (!queue.empty()) {
+      std::unique_ptr<TBufferFile> buffer{queue.front()};
+      files.push_back(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "READ"));
+      queue.pop();
+   }
 
-      fBuffered += buffer->BufferSize();
-      fMerger.AddAdoptFile(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "read"));
+   {
+      std::lock_guard<std::mutex> m(fMergeMutex);
+      for (auto f : files)
+         fMerger.AddAdoptFile(f);
 
-      if (fBuffered > fAutoSave)
-         Merge();
+      fMerger.PartialMerge();
```

Comment:
Do we know if this can take (indirectly) the ROOT global lock?

Reply:
It does take the lock, but I've thought about it and I don't think that this will lead to a dead lock. Otherwise we cannot mix any kind of user lock with any ROOT code. The merge lock can only be aquired by calling the Merge() method (i.e. via the Push() method, which is only called via TBufferMergerFile::Write()). At that point, we do not usually hold a lock on gROOTMutex unless the user has taken the lock himself. So, if we don't have the lock on gROOTMutex, we wait for other ROOT code to finish without problems when we are inside Merge() and call PartialMerge(). If, on the other hand, we do have a lock on gROOTMutex because the user holds it, then taking the lock will have no effect inside Merge(), and the lock will only be released after the merge is finished, so we are again safe. That considered, I think it's not necessary to even move the creation of TMemFiles outside the merge lock at all, so I go back to my previous implementation.

Reply:
> At that point, we do not usually hold a lock on gROOTMutex unless the user has taken the lock himself. 

humm ... so the problem case maybe rare but not impossible.   It should then **at least** be noted that we do **not** support this routine being called while holding the ROOT global write lock. [This would lead to a dead lock if any other thread calls this same routine at about the same time, since that other thread might take the local mutex and then go on to wait on the ROOT global lock while the first thread is waiting on the local mutex].

Reply:
> This would lead to a dead lock if any other thread calls this same routine at about the same time, since that other thread might take the local mutex and then go on to wait on the ROOT global lock while the first thread is waiting on the local mutex.

I don't think that a deadlock is actually possible. If you enter Merge() with the global lock acquired, then you will release the merge lock for sure, since taking the global lock while already having a lock on it is a noop. Then if another thread called Merge() by any means, it will just wait for the first thread to be done merging. If two things call Merge() roughtly at the same time without holding a lock, then one of them will wait for the other, that's the whole point of the merge lock.

Reply:
The deadlock is unlikely but possible:
```
time0, thread 0: grab ROOT global write lock.
time1, thread 1: call TBufferMerger::Merge, grab local mutex, call  fMerger.PartialMerge
time2, thread 1: request ROOT global write lock.
time3, thread 1: is now waiting on thread 0 to release the lock
time4, thread 0: call TBufferMerger::Merge
time5, thread 0: request local mutex
time6, thread 0: is now waiting on thread 1 to release the local mutex.
```
You argue (rightfully) that the user doing
```
time0, thread 0: grab ROOT global write lock.
```
would be odd an rare .. but not impossible.

Reply:
Ah, I see. I did not think about this case. However, if users of ROOT are really allowed to take ROOT's own global lock, the only solution is to go back to using the global mutex around merging. Also, in that case, no code in ROOT can use a lock other than the global lock if it does anything that might take the global lock.

Reply:
In any case, I will merge the first three commits from this branch and think about the last two a bit more.

Reply:
> no code in ROOT can use a lock other than the global lock if it does anything that might take the global lock.

Indeed.  This has pretty much been the strategy (and lead to the 'merging' of the interpreter and ROOT locks).

> the only solution is to go back to using the global mutex around merging. 

There is other options, albeit more complex, for example to make a 'second' concurrent request innocuous and/or ignore and/or just posting for the first one to do more work, etc.  So that the 'Merging' itself can be done without having to use a local locking.

Reply:
> In any case, I will merge the first three commits from this branch and think about the last two a bit more.

Fair enough.  Thanks.